src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -343,53 +206,23 @@ public static IEnumerable<SequenceEdit> GetEdits(SemanticToken[] oldSemanticToke
             }
         }
 
-        /// <summary>
-        /// Stores the values that make up the LSP representation of an individual semantic token.
-        /// </summary>
-#pragma warning disable CA1067 // Override Object.Equals(object) when implementing IEquatable<T>
-        private readonly struct SemanticToken : IEquatable<SemanticToken>
-#pragma warning restore CA1067 // Override Object.Equals(object) when implementing IEquatable<T>
+        // We need to have a shim class because SemanticTokensEdit.Data is an array type, so if we
+        // operate on it directly then every time we append an element we're allocating a new array.
+        private class RoslynSemanticTokensEdit
         {
-            private readonly int _deltaLine;
-            private readonly int _deltaStartCharacter;
-            private readonly int _length;
-            private readonly int _tokenType;
-            private readonly int _tokenModifiers;
-
-            public SemanticToken(int deltaLine, int deltaStartCharacter, int length, int tokenType, int tokenModifiers)
-            {
-                _deltaLine = deltaLine;
-                _deltaStartCharacter = deltaStartCharacter;
-                _length = length;
-                _tokenType = tokenType;
-                _tokenModifiers = tokenModifiers;
-            }
-
-            public void AddToEnd(ArrayBuilder<int> tokensToInsert)
-            {
-                tokensToInsert.Add(_deltaLine);
-                tokensToInsert.Add(_deltaStartCharacter);
-                tokensToInsert.Add(_length);
-                tokensToInsert.Add(_tokenType);
-                tokensToInsert.Add(_tokenModifiers);
-            }
+            public int Start { get; set; }
+            public int DeleteCount { get; set; }
+            public IList<int>? Data { get; set; }
```

Comment:
def doc these fields.

Reply:
can this be a struct instead of a calss?

Reply:
I _think_ unless this makes more sense as a class, since I believe in a struct we can't modify the value of `DeleteCount` without creating an entirely new struct. (If I'm wrong, please let me know. 🙂)